### PR TITLE
feat: add a new range date field component

### PR DESCRIPTION
**Pull request by krozhkov**

### CODE REVIEW

**Comment by korvin89** [comment]
I think that focus should be stand on the first left item in case of click on the right side of the control

https://github.com/gravity-ui/date-components/assets/26501073/de502601-67f8-4d73-bc90-12ecb98f2d61

File: src/components/RangeDateField/hooks/useRangeDateFieldState.ts [hunk 1]
```diff
@@ -0,0 +1,306 @@
+/* eslint-disable complexity */
+import React from 'react';
+
+import {isValid} from '@gravity-ui/date-utils';
+import type {DateTime} from '@gravity-ui/date-utils';
+
+import {
+    type BaseDateFieldState,
+    useBaseDateFieldState,
+} from '../../DateField/hooks/useBaseDateFieldState';
```
Comment (EgorKluch) [inline src/components/RangeDateField/hooks/useRangeDateFieldState.ts:10]: I think it's bad practies to import something from internal component files.
It's better to add hook to component API (index file)
```
import {
    type BaseDateFieldState,
    useBaseDateFieldState,
} from '../../DateField';
```

File: src/components/RangeDateField/utils.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import type {DateTime} from '@gravity-ui/date-utils';
```
Comment (EgorKluch) [inline src/components/RangeDateField/utils.ts:1]: its better to create `utils/getRangeEditableSections.ts` file and add to `utils/index.ts`

File: src/components/RangeDateField/hooks/useRangeDateFieldState.ts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    const sections = useFormatSections(format);
+
+    const allSegments: typeof EDITABLE_SEGMENTS = React.useMemo(() => {
+        return sections
+            .filter((seg) => EDITABLE_SEGMENTS[seg.type])
+            .reduce<typeof EDITABLE_SEGMENTS>((p, seg) => ({...p, [seg.type]: true}), {});
+    }, [sections]);
+
+    // eslint-disable-next-line prefer-const
+    let [validSegments, setValidSegments] = React.useState<RangeValue<typeof EDITABLE_SEGMENTS>>(
```
Comment (korvin89) [inline src/components/RangeDateField/hooks/useRangeDateFieldState.ts:56]: I think it is not a good idea to mutate `validSegments`

Comment (krozhkov) [reply src/components/RangeDateField/hooks/useRangeDateFieldState.ts:56]: Yes. But that's how it works in DateField, and I'm afraid of breaking the logic.

Comment (krozhkov) [reply src/components/RangeDateField/hooks/useRangeDateFieldState.ts:56]: But I'll check it.

File: src/components/RangeDateField/utils/isValidRange.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import {type DateTime, isValid} from '@gravity-ui/date-utils';
+
+import type {RangeValue} from '../../types';
+
+export function isValidRange({start, end}: RangeValue<DateTime>): boolean {
+    return isValid(start) && isValid(end) && start.isBefore(end);
```
Comment (ValeraS) [inline src/components/RangeDateField/utils/isValidRange.ts:6]: Check should be `!end.isBefore(start)`, `start === end` is valid range

File: src/components/RangeDateField/utils/createPlaceholderRangeValue.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import {type PlaceholderValueOptions, createPlaceholderValue} from '../../utils/dates';
+
+export function createPlaceholderRangeValue({placeholderValue, timeZone}: PlaceholderValueOptions) {
+    const start = createPlaceholderValue({placeholderValue, timeZone});
+    const end = start.add(1, 'hour');
```
Comment (ValeraS) [inline src/components/RangeDateField/utils/createPlaceholderRangeValue.ts:5]:
```suggestion
    const end = start;
```

**Review by ValeraS — CHANGES_REQUESTED** [review]

**Review by ValeraS — APPROVED** [review]

**Pull request merged**